rfcs/recursive-type-unrestriction.md

```diff
@@ -45,11 +45,81 @@ Examples like this are quite common in TypeScript code, for example in [ReduxJS]
 
 ## Design
 
-*This section to be filled in once we decide which alternative to use*
+The design is to continue to use strict type aliases, but to use a
+cache during type alias definition. Type aliases are handled as they
+currently are, except for recursive cases.
+
+When defining a type alias `T<a1,...,aN>`, if we encounter a recursive use
+`T<U1,...,UN>` we proceed as follows:
+
+* If every `UI` is `aI`, or if every `UI` does not contain any of the `aJ`s:
```

Comment:
Can we move this to a separate paragraph before the cache description to make it clear that this is the new restriction? The cache then is essentially a description of the mechanics, but from the user perspective the critical part seems to be "either UI = aI for all I, or every UI does not contain any of the aJ/UI for all IJ"